### ISSUE TITLE
Create/edit outline of PDF

### ISSUE THREAD

**Issue by Nokia808** [qpdf/qpdf#343]
Hi. as long as someone started to utilize your QPDF to his GUI (I mean PDF Slicer developer), it will be very useful to add ability for QPDF to be able to manage outline of PDF (adding, creating, editing outline or so called sometime bookmarks) .......

If this functionality added, then it should support levels like this:

chapter one
  section a
  section b
chapter two
  section one
    part one
    part two
  section two
chapter three

& so on ........

The QPDF should be able to label each outline not to specific page only, but to specific height within that page .....

**Comment by jberkenbilt** [comment]
Yes, I agree. I would like to get to this, but I don't expect to get to it any time soon.

**Comment by Nokia808** [comment]
@jberkenbilt 
Hi dear. I wish that you & your family are, currently & for ever, in good health & not affected by COVID19 pandemic ...

Dear I detected a progression in this issue achieved by other program related very closely to your program. It is pikpdf - see the following:
https://github.com/pikepdf/pikepdf/pull/82
the above PR finally merged !
See, kindly, also this commit:
https://github.com/pikepdf/pikepdf/commit/c8459ba9ef466ed32a2d848cbcc9576ec5fb487d

I'm not a programmer & may be wrong. But since pikepdf is very closely related to QPDF it may be useful for this issue related to QPDF itself ......

**Comment by Nokia808** [comment]
@jberkenbilt 
Hi dear. I wish you are & your family in good health from COVID19 pandemic & other diseases.

Is there a hope to add support for this issue in next version ?

Thank you.

**Comment by ifendo** [comment]
Hi, 
just want to check if it is possible to create the bookmarks as we did in Adobe Acrobat when we combine multiple PDF files? 
e.g. combine 
a.pdf
b.pdf
c.pdf
we will create the bookmarks a, b, c by using Adobe Acrobat? I didn't find the option in the manual.

Thank you very much

**Comment by jberkenbilt** [comment]
@ifendo This option is not yet available in qpdf, but it is planned, and I am actively working on changes that will it possible to add. I have no timeline right now, but this is the most desired enhancement at the moment.

**Comment by dbrizard** [comment]
That would be a very interesting feature.
Some small programs already exist, based on `cpdf`, `pdftk` or `gs`, see [here](https://github.com/dbrizard/add-pdf-contents#other-similar-tools), and they take a text file as input to create the outline (depth of indentationgives the level of the bookmark). Maybe some idea are worth reusing...

**Comment by mid-kid** [comment]
I've gone around all of these tools but they all present problems:
- cpdf can't merge outlines of multiple pdfs
- pdftk will "unroll" all outlines, instead of collapsing them
- ghostscript/gs will not preserve 3d objects and has trouble preserving annotations
- mutool merge doesn't preserve 3d objects either
- qpdf cannot merge outlines of multiple pdfs
having a minor breakdown as it's seemingly impossible to find a tool to do this stuff with.

**Comment by jberkenbilt** [comment]
I've been hoping to get to this for years, but I'm actually planning on breaking ground on this work in the next few weeks. It is a complex feature to implement, which is why nothing does it, but hopefully qpdf will be able to do it within the next few months. No promises, but that's my goal.

**Comment by jberkenbilt** [comment]
@mid-kid If you would take a minute to explain what kind of output you would like to see for merged outlines, that would help. I have in mind what I'd like to see, but it would be great to hear from a user who would actually use the feature.

**Comment by mid-kid** [comment]
@jberkenbilt having very little experience with PDFs I'm not entirely sure what to suggest.
I've noticed that some tools also remove the metadata that causes clicking on outlines to zoom into a certain spot of the page, which is annoying.
What I want can be summarized as: join all the outlines from the different PDFs together, updating the page target but preserving as much metadata as possible otherwise. Don't unroll (open) all outlines if they're closed in the source documents, or provide an option to control this. Show the outline when opening the PDF, or provide an option to control this. I know javascript may be a lost cause, but if javascript actions can be preserved as well that'd be amazing.

After a lot of frustration involving PDFs, I ended up coming up with the following script: https://gist.github.com/mid-kid/0cc9f37a7074ad40e65564569a2ba926
Maybe this helps visualize what I mean.

**Comment by jberkenbilt** [comment]
@mid-kid Thanks. I'll come back to this when I get to that point in the pages project, so it will be a while.

**Comment by jberkenbilt** [comment]
This issue is part of the qpdf pages epic. If you are interested in following, please see #1104.

**Comment by foolishgrunt** [comment]
My needs are not as strenuous as those of @mid-kid; all I need is to be able to define bookmarks in the metadata and apply it to a newly-merged PDF file. I wrote a script for this that is almost identical to the script here: https://unix.stackexchange.com/questions/368415/merge-pdf-files-and-automatically-create-a-table-of-contents-with-each-file-as-a/566734#566734.

The main difference in my script, besides being written for Windows Powershell, is that I have a mandate from my boss _not_ to use pdftk for the merge step (that's another story). This makes qpdf the best alternative, and would have been my first choice outright for thr program, except that (to my knowledge) it does not have alternatives to pdftk's "dump_data" and "update_data" commands. This means I use qpdf for all applicable steps _except_ the one where I update the metadata, for which I call pdftk.

It sure would be nice to have a similar function in qpdf!

**Comment by m-holger** [comment]
@foolishgrunt , I am slightly confused by the 'metadata'. Are you basically asking for commands to (1) dump the existing bookmarks to a file (to be used on the individual input files) and (2) replace existing bookmarks with bookmarks uploaded from a file (to be applied to the combined PDF file, using a bookmark file you generate by combining the dumped bookmark files from the induvudual PDF files)?

**Comment by foolishgrunt** [comment]
> @foolishgrunt , I am slightly confused by the 'metadata'. Are you basically asking for commands to (1) dump the existing bookmarks to a file (to be used on the individual input files) and (2) replace existing bookmarks with bookmarks uploaded from a file (to be applied to the combined PDF file, using a bookmark file you generate by combining the dumped bookmark files from the individual PDF files)?

More or less; I'll elaborate on my current process.

When I run
`pdftk \path\to\input.pdf dump_data output \path\to\output.txt`
on a not-yet-bookmarked file, it returns the following text file containing PDF metadata: [metadata.txt](https://github.com/user-attachments/files/16239162/metadata.txt) (Note that for my specific use case, I do not use the `dump_data` command, because none of the files I'm merging have bookmarks already. But such a command could be useful for for other use cases.)

Anyway, to add bookmarks, I generate a new .txt file containing blocks of text following this format:
```
BookmarkBegin
BookmarkTitle: [X]
BookmarkLevel: [Y]
BookmarkPageNumber: [Z]
```
...to be repeated for as many bookmarks as necessary. Here's an example: [bookmarks.txt](https://github.com/user-attachments/files/16239444/bookmarks.txt)

I believe this format for bookmark metadata is customized for use with pdftk, rather than being part of the PDF standard. But it's simple to generate and easy apply to a file:
`pdftk \path\to\input.pdf update_info \path\to\bookmarks.txt output \path\to\bookmarked.pdf`

The bookmark metadata is then appended to the indicated PDF, as seen in this `dump_data` output from the newly bookmarked file: [metadata_bookmarked.txt](https://github.com/user-attachments/files/16239484/metadata_bookmarked.txt)

In case it's helpful, here's the script I've worked out: [script.txt](https://github.com/user-attachments/files/16239710/script.txt)

As I'm writing this, it occurs to me I might be able to hack something together with qpdf using `qpdf --update-from-json`, but... I've never used this command before. It's just something I found in the documentation, and I don't really know its intended purpose. :) I might take a closer look at the two outputs ([metadata_unbookmarked.json](https://github.com/user-attachments/files/16239615/metadata_unbookmarked.json) and [metadata_bookmarked.json](https://github.com/user-attachments/files/16239493/metadata_bookmarked.json)) to see if I can find a useful pattern to exploit one way or another, but this really isn't my forte.

**Comment by m-holger** [comment]
`qpdf --update-from-json` would allow you to this, but it would be pretty complex and I don't think it would be practical.

I have started a discussion about a (future) simpler alternative at #1244.
